GameServerLib/Content/Navigation/NavigationGrid.cs

```diff
@@ -0,0 +1,626 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using GameServerCore;
+using GameServerCore.Content;
+using GameServerCore.Domain.GameObjects;
+using Priority_Queue;
+using LeagueSandbox.GameServer.GameObjects;
+using RoyT.AStar;
+using Vector2 = System.Numerics.Vector2;
+using System.Numerics;
+using GameServerLib.Extensions;
+
+namespace LeagueSandbox.GameServer.Content.Navigation
+{
+    public class NavigationGrid : INavigationGrid
+    {
+        public Vector3 MinGridPosition { get; set; }
+        public Vector3 MaxGridPosition { get; set; }
+        public Vector3 TranslationMaxGridPosition { get; set; }
+        public float CellSize { get; private set; }
+        public uint CellCountX { get; set; }
+        public uint CellCountY { get; set; }
+        public NavigationGridCell[] Cells { get; set; }
+        public ushort[] CellFlags { get; set; } // Version 7 change
+        public Vector2 SampledHeightDistance { get; private set; }
+        public float OffsetX { get; private set; }
+        public float OffsetZ { get; private set; }
+        public float MapWidth { get; private set; }
+        public float MapHeight { get; private set; }
+        public Vector2 MiddleOfMap { get; private set; }
+        public const float SCALE = 2f;
+
+        public NavigationGrid(string fileLocation) : this(File.OpenRead(fileLocation)) { }
+        public NavigationGrid(byte[] buffer) : this(new MemoryStream(buffer)) { }
+        public NavigationGrid(Stream stream)
+        {
+            using (BinaryReader br = new BinaryReader(stream))
+            {
+                byte major = br.ReadByte();
+                ushort minor = major != 2 ? br.ReadUInt16() : (ushort)0;
+
+                this.MinGridPosition = br.ReadVector3();
+                this.MaxGridPosition = br.ReadVector3();
+
+                this.CellSize = br.ReadSingle();
+                this.CellCountX = br.ReadUInt32();
+                this.CellCountY = br.ReadUInt32();
+
+                this.Cells = new NavigationGridCell[this.CellCountX * this.CellCountY];
+                this.CellFlags = new ushort[this.CellCountX * this.CellCountY];
+
+                if (major == 2 || major == 3 || major == 5)
+                {
+                    for (int i = 0; i < this.Cells.Length; i++)
+                    {
+                        this.Cells[i] = new NavigationGridCell(br, i, out this.CellFlags[i]);
```

Comment:
You declared 2 constructors with the same name for different versions. This is very confusing and causes code ambiguity.

Proposed solutions:
Sophisticated solution:
1. Create INavGridCellReader interface with NavGridCell ReadNavGridCell(br, id) function
2. Create abstract class that implements INavGridCellReader and call it NavGridCellReader
3. Create 2 class implementations (for each version) that implement the function ReadNavGridCell
4. Create a static function INavGridCell GetReader(int major) in NavGridCellReader that returns the correct reader for each version
5. Use it as NavGridCellReader.GetReader(major).ReadNavGridCell(br, id)

Simple hotfix solution (not recommended):
1. Add a switch to the constructor for different reading with each version

Reply:
I implemented a solution to this in a commit that's better than creating redundant classes or interfaces

Reply:
The solution isn't really a great one (empty constructor, it should be private at least). I accept the changes.